astropy/modeling/core.py

```diff
@@ -3341,6 +3341,80 @@ def render(self, out=None, coords=None):
         return out
 
 
+    def replace_submodel(self, name=None, model=None):
+        """
+        Construct a new CompoundModel instance from an existing CompoundModel,
+        replacing the named submodel with the new model.
+
+        In order to ensure that inverses and names are kept/reconstructed, it's
+        necessary to rebuild the CompoundModel from the replaced node all the
+        way back to the base. The original CompoundModel is left untouched.
+
+        Parameters
+        ----------
+        model: CompoundModel instance
+            model to be reconstructed with replacement
+        name: str
+            name of submodel to be replaced
+        new_model: Model
+            replacement model
+        """
+        submodels = [m for m in self.traverse_postorder() if m.name == name]
+        if submodels:
+            if len(submodels) > 1:
+                raise ValueError(f"More than one submodel named {name}")
+
+            old_model = submodels.pop()
+            if len(old_model) != len(model):
+                raise ValueError("New and old models must have equal values "
+                                 "for n_inputs")
+
+            # Do this check first in order to raise a more helpful Exception,
+            # although it would fail trying to construct the new CompoundModel
+            if (old_model.n_inputs != model.n_inputs or
+                        old_model.n_outputs != model.n_outputs):
+                raise ValueError("New model must match numbers of inputs and "
+                                 "outputs of existing model")
+
+            tree = _get_submodel_tree(self, name)
+            while tree:
+                branch = self.copy()
+                for node in tree[:-1]:
+                    branch = getattr(branch, node)
+                setattr(branch, tree[-1], model)
+                model = CompoundModel(branch.op, branch.left, branch.right,
+                                      name=branch.name,
+                                      inverse=branch._user_inverse)
+                tree = tree[:-1]
+            return model
+
+        else:
+            raise ValueError(f"No submodels found named {name}")
+
+
+def _get_submodel_tree(model, name, chain=[]):
+    """Find the route down a CompoundModel's tree to the model with the
+    specified name (whether it's a leaf or not)"""
+    links = []
+    try:
+        if model.left.name == name:
+            return chain + ['left']
+        else:
+            links += _get_submodel_tree(model.left, name,
+                                        chain=chain + ['left'])
+    except AttributeError:
+        pass
+    try:
+        if model.right.name == name:
+            return chain + ['right']
+        else:
+            links += _get_submodel_tree(model.right, name,
+                                        chain=chain + ['right'])
+    except AttributeError:
+        pass
+    return links
+
```

Comment:
Does this work when the right operand is combined using fix_inputs? The try clause looks like it would catch it, but it would be good to test against that case.

Reply:
Has `fix_inputs` been fixed? #10170

Reply:
That problem isn't specific to fix_inputs, is it?

Reply:
No, but I'm not sure whether any fix might involve a change to the way `fix_inputs` stores things. I can add a test for my code that (perforce) works with scalars to check it behaves appropriately (I think I tested this locally).

Reply:
`fix_inputs` gave a crash for not having a `name` attribute but the replacement worked. Fixed. Test added.